main/src/library/Iterator.flix

```diff
@@ -14,179 +14,158 @@
+    pub def new(r: Region[r]): Iterator[a, ef, r] \ { Read(r), Write(r) } = Iterator.iterate(r)
 }
 
 ///
 /// Alias for `Foldable.iterator`.
 ///
-pub def iterator(r: Region[r], t: t[a]): Iterator[a, r] \ Write(r) with Foldable[t] =
+pub def iterator(r: Region[r], t: t[a]): Iterator[a, ef, r] \ Write(r) with Foldable[t] =
     Foldable.iterator(r, t)
+*/
 
 namespace Iterator {
 
-    ///
-    /// The type of the done function.
-    ///
-    pub type alias Done[ef: Region] = Unit -> Bool \ ef
+    pub enum Step[a] {
+        case Ans(a)
+        case Skip
+        case Done
+    }
 
     ///
-    /// The type of the next function.
+    /// Returns an iterator built with the stepper function `f`.
     ///
-    pub type alias Next[a: Type, ef: Region] = Unit -> a \ ef
+    pub def iterate(rc: Region[r], f: Unit -> Option[a] \ ef): Iterator[a, ef, r] =
+        let f1 = () -> match f() {
+            case Some(a) => Ans(a)
+            case None    => Done
+        };
+        let iterF = () -> unsafe_cast f1()  as _ \ { ef, r };
+        Iterator(rc, iterF)
 
     ///
     /// Returns an empty iterator.
     ///
-    pub def new(_: Region[r]): Iterator[a, r] \ { Read(r), Write(r) } =
-        let done = () -> unsafe_cast true as _ \ Read(r);
-        let next = () -> unsafe_cast bug!("Empty Iterator!") as _ \ Read(r);
-        unsafe_cast Iterator(done, next) as _ \ { Read(r), Write(r) }
+    pub def empty(rc: Region[r]): Iterator[a, ef, r] =
+        let f = () -> unsafe_cast Done as _ \ ef;
+        Iterator(rc, f)
 
     ///
     /// Returns an iterator containing only a single element, `x`.
     ///
-    pub def singleton(r: Region[r], x: a): Iterator[a, r] \ Write(r) =
-        let d = ref false @ r;
-        let done = () -> deref d;
-        let next = () -> {
-            if (done()) {
-                bug!("Empty Iterator!")
-            }
-            else {
-                d := true;
-                x
-            }
-        };
-        Iterator(done, next)
+    pub def singleton(rc: Region[r], x: a): Iterator[a, ef, r] \ Write(r) =
+        repeat(rc, 1, x)
 
     ///
     /// Returns `Some(x)` if `iter` is not empty. Returns `None` otherwise.
     ///
     /// Consumes the head element of `iter`.
     ///
-    pub def next(iter: Iterator[a, r]): Option[a] \ Read(r) =
-        let Iterator(done, next) = iter;
-        if (done())
-            None
-        else
-            Some(next())
-
-    ///
-    /// Returns true if the iterator is empty.
-    ///
-    /// Does *not* consume any elements of the iterator.
-    ///
-    pub def isEmpty(iter: Iterator[a, r]): Bool \ Read(r) =
-        let Iterator(done, _) = iter;
-        done()
+    pub def next(iter: Iterator[a, ef, r]): Option[a] \ { ef, Read(r) } =
+        let Iterator(_, iterF) = iter;
+        def loop() = match iterF() {
+            case Ans(a) => Some(a)
+            case Skip   => loop()
+            case Done   => None
+        };
+        loop()
```

Comment:
I like it

Reply:
Yes this has worked exactly as hoped - the effect `ef` and region `r` that are encapsulated in the type of the Iterator parameter float to the right of of the type signature. No casting required.